File: 2023/06/haskell/06.hs
{- stack script --resolver lts-14.20 
                --package split -}
module Main where

import Data.List.Split

bb :: Int -> Int -> (Int -> Int) -> Int -> Int
bb l r f d | l == r      = l
           | result <= d = bb (mid + 1) r f d
           | otherwise   = bb l mid f d
  where
    mid    = flip div 2 $ l + r
    result = f mid

calc :: Int -> Int -> Int
calc t x = x * (t - x)

solve :: Int -> Int -> Int
solve t d = t - 2 * x + 1
  where
    x = bb 0 (t `div` 2) (calc t) d

task1 :: [(Int, Int)] -> Int
task1 = product . map (uncurry solve)

task2 :: [(Int, Int)] -> Int
task2 input = uncurry solve input'
  where
    input' = (\(a, b) -> (read a, read b)) $ foldl (\(a, b) (c, d) -> (a ++ c, b ++ d)) ("", "") $ map (\(a, b) -> (show a, show b)) input

main :: IO ()
main = do
  [ts, ds] <- flip fmap getContents $ map (map read . tail . filter (not . null) . splitOn " ") . lines
  let input = zip ts ds
  putStr "Task 1: "
  print $ task1 input
  putStr "Task 2: "
  print $ task2 input

File: 2023/06/rust/06.rs
use std::io;

fn solve(t: &i64, d: &i64) -> i64 {
  let mut l = 0;
  let mut r = t / 2;
  while l < r {
    let m = (l + r) / 2;
    if m * (t - m) <= *d {
      l = m + 1;
    } else {
      r = m;
    }
  }
  t - 2 * l + 1
}

fn task1(input: &Vec<(&i64, &i64)>) -> i64 {
  let mut res = 1;
  for (t, d) in input {
    res *= solve(*t, *d);
  }
  res
}

fn task2(input: &Vec<(&i64, &i64)>) -> i64 {
  let (t2, d2) = input.iter().fold((0, 0), |acc, (t, d)| {
    let (t1, d1) = acc;
    let t2 = (t1.to_string() + &t.to_string()).parse::<i64>().unwrap();
    let d2 = (d1.to_string() + &d.to_string()).parse::<i64>().unwrap();
    (t2, d2)
  });
  solve(&t2, &d2)
}

fn main() {
  let mut line = String::new();

  let _ = io::stdin().read_line(&mut line);
  let ts = line.trim().split(" ").filter(|&x| x != "").skip(1).map(|x| x.parse::<i64>().unwrap()).collect::<Vec<i64>>();
  line.clear();
  let _ = io::stdin().read_line(&mut line);
  let ds = line.trim().split(" ").filter(|&x| x != "").skip(1).map(|x| x.parse::<i64>().unwrap()).collect::<Vec<i64>>();
  let input = ts.iter().zip(ds.iter()).collect::<Vec<(_, _)>>();
  println!("Task 1: {}", task1(&input));
  println!("Task 2: {}", task2(&input));
}
